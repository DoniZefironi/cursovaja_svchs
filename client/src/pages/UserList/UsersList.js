import React, { useContext, useEffect, useState } from 'react';
import { observer } from 'mobx-react-lite';
import { Container, Row, Col, Button } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import './userlist.css';
import { Context } from '../../index';
import Search from '../../components/searcuser';
import Report from '../../components/report';
import UserList from '../../components/userlist';
import UserModal from '../../components/usermodal';
import UserModalCreate from '../../components/usermodalcreate';

const UserContainer = observer(() => {
  const { author, user: currentUser } = useContext(Context); 
  const { users, fetchUsers, createUser, updateUser, currentPage, totalPages } = author;
  const [openIndex, setOpenIndex] = useState(null);
  const [showCreateModal, setShowCreateModal] = useState(false);
  const [showEditModal, setShowEditModal] = useState(false);
  const [currentUserEdit, setCurrentUserEdit] = useState({ id: null, name: '', surname: '', patronymic: '', email: '', phone_number: '', position: '', roles: '' });
  const [newUser, setNewUser] = useState({ name: '', surname: '', patronymic: '', email: '', phone_number: '', position: '', roles: '', password: '' });

  useEffect(() => {
    console.log("Calling fetchUsers...");
    fetchUsers(currentPage);
  }, [fetchUsers, currentPage]);

  useEffect(() => {
    console.log("Users fetched:", users);
  }, [users]);

  const handleToggle = (index) => {
    setOpenIndex(openIndex === index ? null : index);
  };

  const handleCreate = () => {
    createUser(
        newUser.email,
        newUser.password,
        newUser.name,
        newUser.surname,
        newUser.patronymic,
        newUser.phone_number,
        newUser.position,
        newUser.roles
    );
    setShowCreateModal(false);
    setNewUser({ name: '', surname: '', patronymic: '', email: '', phone_number: '', position: '', roles: '', password: '' });
    window.location.reload();  
};


const handleEdit = () => {
    updateUser(
        currentUserEdit.id,
        currentUserEdit.name,
        currentUserEdit.surname,
        currentUserEdit.patronymic,  
        currentUserEdit.email,
        currentUserEdit.phone_number,
        currentUserEdit.position,
        currentUserEdit.roles
    );
    setShowEditModal(false);
    setCurrentUserEdit({ id: null, name: '', surname: '', patronymic: '', email: '', phone_number: '', position: '', roles: '' });
    window.location.reload(); 
};

  const handleChange = (e, key, isNew = false) => {
    const value = e.target.value;
    if (isNew) {
      setNewUser({ ...newUser, [key]: value });
    } else {
      setCurrentUserEdit({ ...currentUserEdit, [key]: value });
    }
  };

  const handlePageChange = (newPage) => {
    if (newPage > 0 && newPage <= totalPages) {
      fetchUsers(newPage);
    }
  };

  return (
    <div className='main'>
      <Container className="mt-4">
        <Row className='gapchek'>
          <Col md={8}>
            <UserList
              users={users}
              currentUser={currentUser}
              handleToggle={handleToggle}
              openIndex={openIndex}
              setCurrentUserEdit={setCurrentUserEdit}
              setShowEditModal={setShowEditModal}
            />
            <div className="pagination-controls mt-3">
              <Button
              style={{backgroundColor : "#2B579A"}}
                onClick={() => handlePageChange(currentPage - 1)}
                disabled={currentPage === 1}
              >
                пред.
              </Button>
              <span>{`Page ${currentPage} of ${totalPages}`}</span>
              <Button
              style={{backgroundColor : "#2B579A"}}
                onClick={() => handlePageChange(currentPage + 1)}
                disabled={currentPage === totalPages}
              >
                след.
              </Button>
            </div>
          </Col>
          <Col md={4}>
            <div className="content-box filter-box">
              <Search />
              {currentUser.user.roles.includes("ADMIN") && ( 
                <>
                <h5>Создать преподавателя</h5>
                          <Button variant="light" onClick={() => setShowCreateModal(true)}>Создать</Button>
                </>
        )}
            </div>
          </Col>
        </Row>
      </Container>

      <UserModalCreate
        show={showCreateModal}
        onHide={() => setShowCreateModal(false)}
        user={newUser}
        handleChange={(e, key) => handleChange(e, key, true)}
        handleSave={handleCreate}
        title="Создать новую запись"
      />

      <UserModal
        show={showEditModal}
        onHide={() => setShowEditModal(false)}
        user={currentUserEdit}
        handleChange={(e, key) => handleChange(e, key)}
        handleSave={handleEdit}
        title="Редактировать запись"
      />
    </div>
  );
});

export default UserContainer;
