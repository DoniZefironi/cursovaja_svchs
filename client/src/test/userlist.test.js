import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import UserList from '../components/userlist'; // Убедитесь, что путь правильный

const users = [
  { surname: 'Ivanov', name: 'Ivan', patronymic: 'Ivanovich', position: 'Engineer', email: 'ivan@example.com', phone_number: '123456789', roles: ['USER'] },
  { surname: 'Petrov', name: 'Petr', patronymic: 'Petrovich', position: 'Manager', email: 'petr@example.com', phone_number: '987654321', roles: ['ADMIN'] }
];

const currentUser = { user: { roles: ['ADMIN'] } };

const handleToggle = jest.fn();
const setCurrentUserEdit = jest.fn();
const setShowEditModal = jest.fn();

describe('UserList Component', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(
      <UserList
        users={users}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentUserEdit={setCurrentUserEdit}
        setShowEditModal={setShowEditModal}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders with correct user information', () => {
    const { getByText, queryByText } = render(
      <UserList
        users={users}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentUserEdit={setCurrentUserEdit}
        setShowEditModal={setShowEditModal}
      />
    );

    // Проверка наличия пользователей
    expect(getByText('Ivanov Ivan Ivanovich')).toBeInTheDocument();
    expect(getByText('Должность: Engineer')).toBeInTheDocument();
    expect(getByText('Petrov Petr Petrovich')).toBeInTheDocument();
    expect(getByText('Должность: Manager')).toBeInTheDocument();

    // Проверка наличия кнопок
    expect(queryByText('No users available')).not.toBeInTheDocument();
  });

  it('handles toggle correctly', () => {
    const { getAllByRole } = render(
      <UserList
        users={users}
        currentUser={currentUser}
        handleToggle={handleToggle}
        openIndex={null}
        setCurrentUserEdit={setCurrentUserEdit}
        setShowEditModal={setShowEditModal}
      />
    );

    // Проверка вызова handleToggle
    const buttons = getAllByRole('button');
    fireEvent.click(buttons[0]);
    expect(handleToggle).toHaveBeenCalledWith(0);
  });

  it('shows and hides additional information on toggle', () => {
    const { getByText, queryByText, rerender } = render(
        <UserList
            users={users}
            currentUser={currentUser}
            handleToggle={handleToggle}
            openIndex={-1} // Начальное скрытое состояние
            setCurrentUserEdit={setCurrentUserEdit}
            setShowEditModal={setShowEditModal}
        />
    );

    // Проверка начального состояния
    expect(queryByText('Email: ivan@example.com')).not.toBeVisible();
    expect(queryByText('Email: petr@example.com')).not.toBeInTheDocument();

    // Обновляем рендеринг компонента с openIndex = 0
    rerender(
        <UserList
            users={users}
            currentUser={currentUser}
            handleToggle={handleToggle}
            openIndex={0} // Открываем первый элемент
            setCurrentUserEdit={setCurrentUserEdit}
            setShowEditModal={setShowEditModal}
        />
    );

    // Проверка состояния после обновления
    expect(getByText('Email: ivan@example.com')).toBeInTheDocument();
    expect(queryByText('Email: petr@example.com')).not.toBeInTheDocument();

    // Скрываем информацию снова
    rerender(
        <UserList
            users={users}
            currentUser={currentUser}
            handleToggle={handleToggle}
            openIndex={-1} // Скрываем информацию
            setCurrentUserEdit={setCurrentUserEdit}
            setShowEditModal={setShowEditModal}
        />
    );

    // Проверка состояния после скрытия
    expect(queryByText('Email: ivan@example.com')).not.toBeInTheDocument();
    expect(queryByText('Email: petr@example.com')).not.toBeInTheDocument();
});
});
