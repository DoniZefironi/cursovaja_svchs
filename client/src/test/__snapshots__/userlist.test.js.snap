// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="content-box list-box width100"
  >
    <div
      class="contentiks"
    >
      <div
        class="list-item"
      >
        <div>
          <h6>
            Ivanov Ivan Ivanovich
          </h6>
          <p>
            Должность: Engineer
          </p>
        </div>
        <div
          class="iconochke"
        >
          <button
            class="plus-button btn btn-light"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
          </button>
          <button
            class="btn btn-warning"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="additional-info collapse"
      >
        <p>
          Email: ivan@example.com
        </p>
        <p>
          Phone: 123456789
        </p>
        <p>
          Role: USER
        </p>
      </div>
    </div>
    <div
      class="contentiks"
    >
      <div
        class="list-item"
      >
        <div>
          <h6>
            Petrov Petr Petrovich
          </h6>
          <p>
            Должность: Manager
          </p>
        </div>
        <div
          class="iconochke"
        >
          <button
            class="plus-button btn btn-light"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
              />
            </svg>
          </button>
          <button
            class="btn btn-warning"
            type="button"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 576 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M402.6 83.2l90.2 90.2c3.8 3.8 3.8 10 0 13.8L274.4 405.6l-92.8 10.3c-12.4 1.4-22.9-9.1-21.5-21.5l10.3-92.8L388.8 83.2c3.8-3.8 10-3.8 13.8 0zm162-22.9l-48.8-48.8c-15.2-15.2-39.9-15.2-55.2 0l-35.4 35.4c-3.8 3.8-3.8 10 0 13.8l90.2 90.2c3.8 3.8 10 3.8 13.8 0l35.4-35.4c15.2-15.3 15.2-40 0-55.2zM384 346.2V448H64V128h229.8c3.2 0 6.2-1.3 8.5-3.5l40-40c7.6-7.6 2.2-20.5-8.5-20.5H48C21.5 64 0 85.5 0 112v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V306.2c0-10.7-12.9-16-20.5-8.5l-40 40c-2.2 2.3-3.5 5.3-3.5 8.5z"
              />
            </svg>
          </button>
        </div>
      </div>
      <div
        class="additional-info collapse"
      >
        <p>
          Email: petr@example.com
        </p>
        <p>
          Phone: 987654321
        </p>
        <p>
          Role: ADMIN
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
